std-bits/table/src/main/java/org/enso/table/format/xlsx/Reader.java

```diff
@@ -0,0 +1,182 @@
+      throws IOException {
+    return read_table(new XSSFWorkbook(is), sheetIdx, hasHeaders, unnamedColumnPrefix, mkDate);
+  }
+
+  /**
+   * Reads the specified XLS file into a table.
+   *
+   * @param is an {@link InputStream} allowing to read the XLS file contents.
+   * @param sheetIdx specifies which sheet should be read. If the value is a {@link Long}, it is
+   *     used as a 0-based index of the sheet. If it is a {@link String}, it is used as a sheet
+   *     name. Otherwise, the active sheet is read.
+   * @param hasHeaders specifies whether the first non-empty row of the sheet should be used for
+   *     column names.
+   * @param unnamedColumnPrefix specifies the prefix to use for missing columns.
+   * @param mkDate a function converting java-based dates into a format understandable by the
+   *     caller.
+   * @return a {@link Table} containing the specified data.
+   * @throws IOException when the input stream cannot be read.
+   */
+  public static Table read_xls(
+      InputStream is,
+      Object sheetIdx,
+      boolean hasHeaders,
+      String unnamedColumnPrefix,
+      Function<LocalDate, Value> mkDate)
+      throws IOException {
+    return read_table(new HSSFWorkbook(is), sheetIdx, hasHeaders, unnamedColumnPrefix, mkDate);
+  }
+
+  private static Table read_table(
+      Workbook workbook,
+      Object sheetIdx,
+      boolean hasHeaders,
+      String unnamedColumnPrefix,
+      Function<LocalDate, Value> mkDate)
+      throws IOException {
+    Sheet sheet = null;
+    if (sheetIdx instanceof Long) {
+      sheet = workbook.getSheetAt(((Long) sheetIdx).intValue());
+    } else if (sheetIdx instanceof String) {
+      sheet = workbook.getSheet((String) sheetIdx);
+    }
+    if (sheet == null) {
+      sheet = workbook.getSheetAt(workbook.getActiveSheetIndex());
+    }
+    int minRow = sheet.getFirstRowNum();
+    int maxRow = sheet.getLastRowNum();
+    if (minRow == -1) {
+      return new Table(new Column[0]);
+    }
+    int minCol = Integer.MAX_VALUE;
+    int maxCol = -1;
+    for (int i = minRow; i <= maxRow; i++) {
+      var row = sheet.getRow(i);
+      if (row == null) continue;
+      var firstCell = row.getFirstCellNum();
+      var lastCell = row.getLastCellNum();
+      if (firstCell != -1 && lastCell != -1) {
+        minCol = Math.min(minCol, firstCell);
+        maxCol = Math.max(maxCol, lastCell);
+      }
+    }
+    if (minCol >= maxCol) {
+      return new Table(new Column[0]);
+    }
+
+    List<String> colNames = new ArrayList<>(maxCol - minCol);
+    if (hasHeaders) {
+      var headerRow = sheet.getRow(minRow);
+      minRow++;
+      for (int i = minCol; i < maxCol; i++) {
+        var cell = headerRow.getCell(i, Row.MissingCellPolicy.RETURN_BLANK_AS_NULL);
+        if (cell == null || cell.getCellType() != CellType.STRING) {
+          colNames.add(unnamedColumnPrefix + (i - minCol));
+        } else {
+          colNames.add(cell.getStringCellValue());
+        }
+      }
```

Comment:
I think our dataframes do not deal well with non-unique column names, so unless we _do want_ to support this (in which case it would be good to create an issue tracking that, because currently many operations can lead to undefined behaviour in presence of name clashes in table columns), I think we should do something here to deduplicate these column names.

Reply:
I'm not sure what is the case with CSV loading, but if it is the same thing maybe it should also get some deduplication? Or at least some kind of a warning (although that will not show up in the IDE, at least for now). Because I think this can lead to some very confusing situations and I'm sure there are CSVs/Excel sheets with duplicates exist in the wild.